rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,491 @@
+the graph def that contains the new to process the new version, even though the
+only thing that *has* changed is the tensorflow added a faster implementation of
+that Op.
+
+For example: consider matrix exponentiation. It is implemented as a fairly
+complex tensorflow graph that uses just regular matrix multiplication and other
+standard ops. However, one can easily imagine that this implementation could be
+highly optimized if done as a single expm node, however if we replace that, it
+will break old tools, which would either need to effectively re-implement
+tensorflow original implementation.
+
+Requires custom version of tensorflow to add new ops outside of tensorflow which
+makes it out of reach for most users and introduces incompatibilities within the
+ecosystem, effectively forcing large users to be stuck at old versions of tensorflow.
+
+#### Alternative 2: Use name scopes for delineation of the ops
+
+Pros:
+
+*   Simple and intuitive graph-def
+*   Backward compatible - no new operations are added.
+
+Cons:
+
+*   very brittle for the tools to optimize such ops. If they depend on scope
+    names it can easily cause conflicts with models that are doing something
+    else or accidental renames can cause scopes to become invisible for the
+    graph processing.
+*   If tools depend on graph pattern matching, this makes it hard to change
+    implementations later on.
+
+*   Tooling is not forward compatible.
+
+##
+
+## Appendix: Future support for optimizing gradient functions
+
+This proposal is concerned with the inference of the composite ops, assuming that the downstream tooling today 
+rarely if ever deals with gradients, and when it does, it can just rely on the provided implementation. However
+eventually this is likely to change, and thus we would like to keep the door open to extending this RFC to support 
+optimization of both foward and backward passes.  In this appendix we provide several  options of how 
+this could be potentially supported in the future, for the reference purposes.  
+
+
+### Background: why is it hard
+Suppose we have implemented function `my_function(x) = very(complex(implementation(x)))`. 
+Now, if some downstream library would like to support optimized implemetnation of `my_function`,
+all it needs to do is 
+to replace the tensorflow implementation with its own implementation. However, if at 
+any point we need to compute gradient of `my_function(x)`, then to avoid recomputation, 
+the default implementation of gradient would need all intermediate values produced 
+by tensorflow implementation. In this case this would be values of 
+`implementation(x)`, `complex(implementation(x))`, etc. 
+
+This is problematic for two reasons: 
+
+  1) Downstream have dependence on tensorflow's implementation of the composite ops, and for instance can't just choose arbitrary implementation
+  2)  If tensorflow implementation changes so needs downstream's.
```

Comment:
Nit: you likely should go through a reformatter for the document, some paragraphs need reflowing and some lines are very long.

Reply:
Done